### PR TITLE
Allow forced matching on pillar top

### LINKED ISSUE #27709: Pin minions to pillar environments
I can't figure out how to do this, so I'm listing this as a "It'd be nice to have..".

We plan on having several organizations using the same salt-master to do their work, but within a multi-tenant environment, we need to make sure that groupA isn't affecting groupB's servers.

Some of this is ACLs and which hosts certain groups can issue commands on, but other parts ( namely pillars ) don't seem to have any 'walls' that we can slide in place.

We can certainly do something like:

```
pillar_roots:
  - groupA:
    - /srv/pillar/groupA
  - groupB:
    - /srv/pillar/groupB
```

And then allow groupA access to the /srv/pillar/groupA folder, same for groupB.

But, it seems with pillar data, if some engineer accidentally ends up targeting '*' in any of those pillar SLS files in his/her environment, it affects every minion, even minions that are environment pinned in the minion config.  Obviously, there are some other ways to try and negate this ( education, linting, code reviews or an ext_pillar that handles it.. ), but it'd be nice if there was some way to restrict the pillar data a minion got based on environment.  It seems the file_root stuff has this capability, but not the pillar.

### PR DESCRIPTION
**Pull request by sjmh**
This allows for forcing a compound match on a pillar env in an attempt to make top-file modifications to be 'safer' in an attempt to fix #27709.

For every pillarenv, you can specify a list of matchers.  If the saltenv being matched against is in the pillarenv_force_match dict, it will add the matchers to the matcher specified in the top file in that environment.

If '*' is used in the top file of an environment and the environment is in the pillarenv_force_match, it just discards the '*' and uses the matcher from pillarenv_force_match.

So, if we have something like this ( ignoring the fact that matching on grains is insecure, it works better as an example in my lab since I don't have a huge diverse set of hosts ):

```
pillar_roots:
  base:
    - /srv/pillar
  ctg:
    - /srv/pillar/ctg
  sbg:
    - /srv/pillar/sbg

pillarenv_force_match:
  ctg:
    - 'G@bu:ctg'
  sbg:
    - 'G@bu:sbg'
```

And we have this in our /srv/pillar/top.sls:

```
base:
  '*':
    - core
```

And this in /srv/pillar/ctg/top.sls:

```
ctg:
  '*':
    - ctg_core
  'G@role:db':
    - ctg_db
```

All servers would get core. ctg_core gets applied only to servers that match G@bu:ctg.  ctg_db only gets applied to servers that match 'G@role:db AND G@bu:ctg'.  

I haven't tested this with the git_pillar but the hope was that I could combine this with git_pillar and allow other organizations to control their own repo and top file within that, without me having to worry about them accidentally clobbering other org's pillar data, or leaking their pillar data to a host they accidentally target ( or globally.. ), because I've effectively limited their targeting to a subset of hosts.

I'm not sure this is ready for a merge, but I wanted to get some feedback on it and if this would be the right area to implement something like this.

### CODE REVIEW

**Comment by cachedout** [comment]
I totally see the use case for this. What you're identifying here, I think, is a clear shortcoming in the segmentation of pillar among disparate teams in a large organization. If I'm understanding you correctly, you want to be able to say, `OK Team Y, here's your top file and don't worry, you won't ever be able to step on the toes of minions belonging to Team X`. If I'm misunderstanding, please let me know.

That said, I haven't looked closely at the implementation but this feels effectively like an ACL that merges onto top files. Is that a fair estimation of where you're going with this? Typically, I don't really like multiple sources of truth for things like this, but in this particular case you may have identified a completely sane use case.

I know that's not exactly well-targeted feedback but I just wanted to move the discussion forward and make sure I understand what you're looking to achieve here. Interesting stuff.

**Comment by sjmh** [comment]
@cachedout - That's it exactly.  Because of political and historical reasons, we can't really use syndics for business-unit separation, so we're in a situation where we have multiple organizations using the same master.  We want to have some restrictions to targeting pillar, while at the same time, allowing users the freedom to manage the targeting for the systems they manage, rather than forcing us to be a gatekeeper of a global top file.

**Comment by whiteinge** [comment]
I'll second what's been said thus far: extremely valid use case. This is a fairly common ask for larger team environments, in my experience. And it's one Salt should add a better answer to.

I'm a bit uncomfortable with spreading the matching over the two settings plus the main top file though. It took me a few minutes of reading between them before I realized the intent.

I have two thoughts:

1. The code implementation is simple but I think it's still worth exploring different, simpler config patterns.
2. Could this be done with existing Salt using Jinja to dynamically build the top file? (This is more of an immediate fix while we stew on the first one more.)

Do you think something like the following could get you close enough to satisfy the separate top files requirement? (Note, not fully tested.)

```yaml
# /srv/pillar/ctg/top.sls
ctg:
  '*':
    - ctg_core
  'G@role:db':
    - ctg_db


# /srv/pillar/top.sls
{% load_yaml as common %}
base:
  '*':
    - core
{% endload %}

{% set current_team = salt.match.filter_by({
  'G@bu:ctg': 'ctg',
  'G@bu:sbg': 'sbg',
}) %}

{% import_yaml salt.file.join('/srv/pillar', current_team, 'top.sls') as team_top %}

# Recursive merge, appends list items. See pull req: #30532
# Otherwise {{ common.merge(team_top) | json() }} will work OOB if there is no key overlap.
{{ salt.slsutil.update(common, team_top, merge_lists=True) | json() }}

# Result is:
# base:
#   '*':
#     - core
# ctg:
#   '*':
#     - ctg_core
#   'G@role:db':
#     - ctg_db
```

**Comment by thatch45** [comment]
The implementation logic here is also sound, I think this is a great feature to add to Salt.
@sjmh can you please add documentation and tests on this? This is a core security feature and will require tests before we can merge it since a regression here could result in a CVE. Also, an undocumented feature is a non-existent feature.
Thanks for the addition @sjmh ! Let us know if you need help with knowing where to document or help with adding tests!

**Comment by sjmh** [comment]
Yeah, I agree that moving this out into the config, and as a second option, isn't the cleanest way of doing this. 

@whiteinge - unless I'm misunderstanding the end result in your example, that wouldn't really be what we were aiming for.  The goal is for something like:

/srv/pillar/ctg/top.sls
```
ctg:
  '*':
    - ctg_core
  'G@role:db':
    - ctg_db
```

/srv/pillar/top.sls
```
base:
  '*':
    - core
```

To merge into:
```
base:
  '*':
    - core
ctg:
  'G@bu:ctg':
    - ctg_core
  'G@bu:ctg and G@role:db':
    - ctg_db
```

In your example, the ctg/top.sls is still pushing out ctg_core to all hosts and that's something we don't want organizations to be able to do.

Ideally, I think this would probably be better as a function of a top file merge strategy.  One of the main benefits being that any changes to the top-level filtering in the main top.sls wouldn't require a restart of the master.  Something like:

/srv/pillar/top.sls:
```
base:
  '*':
    - core
ctg:
  load: /srv/pillar/ctg/top.sls
  match: 'G@bu:ctg'
sbg:
  load: /srv/pillar/sbg/top.sls
  match: 'G@bu:sbg'
```

Another idea might be to include it in the pillar_roots config declaration.  That way, you're declaring the environments and the forced-match at the same time.  Unfortunately, I can't think of a way that makes sense to do that without breaking backwards compatibility with pillar_roots.

**Comment by whiteinge** [comment]
> In your example, the ctg/top.sls is still pushing out ctg_core to all hosts

Whoops, missed that. You're quite right. For that I'd grab the target from the lookup table and do a replace after the merge.

That said, @thatch45 gave your OP approach the :+1: so nevermind alternate ideas. :smile: This'll be a nice feature addition.

**Comment by cachedout** [comment]
Indeed. I think we're fairly settled on a direction here. We'll await docs and tests unless there are more issues to be resolved. Good discussion all, and thanks very much @sjmh

**Comment by basepi** [comment]
:+1: this will be awesome.

**Comment by sjmh** [comment]
Thanks for the discussion guys.  I'll work on getting the documentation and tests in this weekend/week.  

As an off-topic aside, I'll be leaving on paternity leave in about 2 weeks ( with twins ), so if this doesn't see action by then, I've been sucked in the world of parenting and probably wouldn't be able to finish this off till the beginning of March.

**Comment by s0undt3ch** [comment]
@sjmh there's at least two Dad's of twins in the SaltStack company, I'm one of them. We'll tottaly understand if you fall off the radar for some months :smile:

Congratulations!!!! And enjoy them...

 Don't worry, the world won't end(you'll understand in a couple of weeks why I said this), and there's nothing better than Them!

**Comment by s0undt3ch** [comment]
Oh! Oh! Sleep! Every minute that you have a chance to...

Again, Congratulations!

**Comment by sjmh** [comment]
I've added a new change to this as well that should probably be looked at.  I found ( while trying to break my first change ), that it was somewhat easy to by-pass the restriction the force-match was trying to enforce.  And, again, we're really trying to say, as @cachedout put it so elegantly: ```OK Team Y, here's your top file and don't worry, you won't ever be able to step on the toes of minions belonging to Team X```.  But, turns out you still can.

If you have:

/etc/salt/master
```
pillar_roots:
  base:
    - /srv/pillar
  ctg:
    - /srv/pillar/ctg
  sbg:
    - /srv/pillar/sbg

pillarenv_force_match:
  ctg:
    - 'G@organization:ctg'
```

/srv/pillar/top.sls
```
base:
  '*':
    - corestuff
```

The /srv/pillar/ctg/top.sls could be formatted like so:
```
base:
  '*':
    - ctg.ctgcorestuff
```

Which would enable you to skip the forced-match and apply all of your pillar files to all hosts ( or whatever other match you came up with ).

The additional change goes through the resulting top dict after all the top files have been compiled ( but not merged ).  If it finds that the environment is in the pillarenv_force_match, it only allows definitions from that top file that match the environment name.  In the example above, the 'base' stanza from the /srv/pillar/ctg/top.sls would be discarded.

The whole goal being able to tell teams, with pretty good confidence, that they don't have to worry about stepping on the toes of other teams and vice versa.

Thoughts?

File: salt/pillar/__init__.py [hunk 1]
```diff
@@ -440,6 +448,12 @@ def top_matches(self, top):
                 if saltenv != self.opts['pillarenv']:
                     continue
             for match, data in six.iteritems(body):
+                if saltenv in self.opts['pillarenv_force_match'].keys():
```
Comment (thatch45) [inline salt/pillar/__init__.py]: Do we need the .keys() here?

**Comment by thatch45** [comment]
Very nice catch @sjmh ! I agree, we need to make sure that this is isolated, at least from non nefarious intent.

Speaking of which, we need to add this option to the docs, and when we do, we need to specify that this is intended for team workflow isolation and should not be considered an ironclad security isolation mechanism, just so that we don't have to CVE bugs in here :)

**Comment by cachedout** [comment]
Bump @sjmh 

Just making sure you saw the comments from @thatch45 above. Thanks!

**Comment by sjmh** [comment]
Yup, saw.  Twins were born last Monday so been in full on kid mode :). Will get these changes/docs in as soon as I'm able.

**Comment by cachedout** [comment]
My fault! Totally forgot about that. Congratulations! No hurry of course. Enjoy your time with the little ones. :]

**Comment by cachedout** [comment]
Hi again @sjmh 

I'm sure you're still busy with your kids but I just wanted to check in with you here. Thanks!

**Comment by sjmh** [comment]
Hey @cachedout - thanks for the check in :) I'm hoping to have this completed by end of next week.

**Comment by cachedout** [comment]
@sjmh Great! Thank you very much.

**Comment by sjmh** [comment]
Added the necessary documentation bits if you'd like to review those.  Working on tests this week.

File: conf/master [hunk 2]
```diff
@@ -624,6 +624,7 @@
 # Recursively merge lists by aggregating them instead of replacing them.
 #pillar_merge_lists: False
 
+<<<<<<< HEAD
```
Comment (cachedout) [inline conf/master]: Merge conflict snuck in. :]

File: conf/master [hunk 3]
```diff
@@ -661,6 +662,19 @@
...
+# matchers defined that are defined within an environment. This helps prevent
+# pillar data from leaking or being overridden between environments.  This
+# feature cannot fully prevent pillar data leakage or overwrites and should
+# not be used as a security measure.
+#pillarenv_force_match:
+#  dev:
+#    - '*dev*'
+#  staging:
+#    - 'G@role:staging'
+>>>>>>> Add documentation for pillarenv_force_match
```
Comment (cachedout) [inline conf/master]: Another merge conflict marker here

File: conf/master (hunk 2 above)
Comment (sjmh) [inline conf/master]: Oops! Will take care of that.  It was late.  I was tired.  Twins.  Excuses. :)

Comment (cachedout) [inline conf/master]: Twins will do that. Just ask @terminalmage or @s0undt3ch ;]

**Comment by cachedout** [comment]
@sjmh Just checking in on this one today. Let me know how we're doing. Thanks!

**Comment by cachedout** [comment]
@sjmh Hi again. :] Just making the rounds and seeing where we are here.

**Comment by sjmh** [comment]
Hey @cachedout - I've got some preliminary stuff done for the tests.  Unless I'm missing where the tests are, none of the functions in _init_.py have any tests, and the Pillar init function is rather heavy, so been trying to create the tests in a way that they can be used easily for the other functions.

**Comment by cachedout** [comment]
Hi @sjmh I'm not too surprised that stuff isn't under test, unfortunately. Thanks for taking this on. It's much appreciated!

**Comment by cachedout** [comment]
Hi @sjmh Just checking in on this. Thanks!

**Comment by cachedout** [comment]
Hi @sjmh 

I'm sure things are busy over there. Since I haven't heard back from you, I'm going to close this for now. Of course, we still want it when you're ready to return to it so feel free to re-submit it when it's ready. Thanks!
